File: app/src/main/java/com/nlnd/moodler/WatchMoodle.java
package com.nlnd.moodler;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import processing.android.CompatUtils;
import processing.android.PFragment;
import processing.core.PApplet;

public class WatchMoodle extends AppCompatActivity
{
    Context context;
    Uri file;
    private static Boolean render;
    Boolean loaded;
    private static int duration;
    private Uri musicFile;
    private static String[] config;
    private static MediaPlayer mplayer;

    @Override
    public void onSaveInstanceState(Bundle outState)
    {
        outState.putParcelable("file", file);
        outState.putBoolean("render", render);
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        Uri ownFile;
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);

        FrameLayout frame = new FrameLayout(this);
        context = this;

        frame.setId(CompatUtils.getUniqueViewId());
        setContentView(frame, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT));
        if (savedInstanceState == null)
        {
            file = Objects.requireNonNull(getIntent().getExtras()).getParcelable("file");
            render = getIntent().getExtras().getBoolean("render");
        }
        else
        {
            file = savedInstanceState.getParcelable("file");
            render = savedInstanceState.getBoolean("render");
        }

        ownFile = file;
        try
        {
            loadFile(new File(ownFile.getPath()));
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
        mplayer = new MediaPlayer();
        mplayer.setAudioStreamType(AudioManager.STREAM_MUSIC);

        try
        {
            mplayer.setDataSource(this, musicFile);
            mplayer.prepare();
            duration = mplayer.getDuration();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        PApplet sketch = new WatchMoodleSketch();
        PFragment fragment = new PFragment(sketch);
        fragment.setView(frame, this);
    }

    private void loadFile(File file) throws Exception
    {
        List<String> ar = new ArrayList<>();
        DataInputStream dis = new DataInputStream(new FileInputStream(file));
        DataOutputStream dos;
        String newSong = (getFilesDir().toString() + "/tempSong.mp3");
        String line;

        while (true)
        {
            line = dis.readLine();
            ar.add(line);
            if (line.equals("end"))
                break;
        }
        config = new String[ar.size()];
        config = ar.toArray(config);

        byte[] b = new byte[dis.available()];
        dos = new DataOutputStream(new FileOutputStream(newSong));
        dis.read(b);
        dos.write(b);

        musicFile = Uri.fromFile(new File(newSong));
        loaded = true;
    }

    public static void stopMusic()
    {
        mplayer.stop();
    }

    public static int getCurrentPosition()
    {
        return mplayer.getCurrentPosition();
    }

    public static void pauseMusic()
    {
        mplayer.pause();
    }

    public static void startMusic()
    {
        mplayer.start();
    }

    public static void musicSeekTo(int a)
    {
        mplayer.seekTo(a);
    }

    public static void mute()
    {
        mplayer.setVolume(0f, 0f);
    }

    public static Boolean getRender()
    {
        return render;
    }

    public static String[] getConfig()
    {
        return config;
    }

    public static MediaPlayer getMplayer()
    {
        return mplayer;
    }

    public static int getDuration()
    {
        return duration;
    }
}

File: feature/src/main/java/com/nlnd/moodler/feature/WatchMoodle.kt
package com.nlnd.moodler.feature

import android.app.Activity
import android.content.Intent
import android.media.AudioManager
import android.media.MediaPlayer
import android.net.Uri
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.FrameLayout
import processing.android.CompatUtils
import processing.android.PFragment
import processing.core.PApplet
import java.io.*

class WatchMoodle : AppCompatActivity()
{
	private var sketch : PApplet? = null
	private var loaded = false

	override fun onCreate(savedInstanceState: Bundle?)
	{
		var ownFile : Uri?
		super.onCreate(savedInstanceState)
		requestWindowFeature(Window.FEATURE_NO_TITLE)
		getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
				WindowManager.LayoutParams.FLAG_FULLSCREEN)

		val frame = FrameLayout(this)
		context = this

		frame.id = CompatUtils.getUniqueViewId()
		setContentView(frame, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
				ViewGroup.LayoutParams.MATCH_PARENT))
		if (savedInstanceState == null)
		{
			file = intent.extras.getParcelable("file")
			render = intent.extras.getBoolean("render")
		}

		if (file == null)
			return

		ownFile = file
		loadFile(File(ownFile!!.path))

		while(!loaded)
			continue

		mplayer = MediaPlayer()
		mplayer!!.setAudioStreamType(AudioManager.STREAM_MUSIC)

		try
		{
			mplayer!!.setDataSource(this, musicFile)
			mplayer!!.prepare()
			duration = mplayer!!.duration
		}
		catch (e: IOException)
		{
			e.printStackTrace()
		}

		sketch = WatchMoodleSketch()
		val fragment = PFragment(sketch)
		fragment.setView(frame, this)
	}

	override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray)
	{
		if (sketch != null)
		{
			sketch!!.onRequestPermissionsResult(
					requestCode, permissions, grantResults)
		}
	}

	override fun onNewIntent(intent: Intent)
	{
		if (sketch != null)
		{
			sketch!!.onNewIntent(intent)
		}
	}

	fun loadFile(f : File)
	{
		val ar = ArrayList<String>()
		val dis = DataInputStream(FileInputStream(f))
		val dos: DataOutputStream
		val newSong = (filesDir.toString() + "/tempSong.mp3")
		var line : String

		while (true)
		{
			line = dis.readLine()
			ar.add(line)
			if (line == "end")
				break
		}
		config = ar.toTypedArray()

		val b = ByteArray(dis.available())
		dos = DataOutputStream(FileOutputStream(newSong))
		dis.read(b)
		dos.write(b)

		musicFile = Uri.fromFile(File(newSong))
		loaded = true
	}

	override fun onDestroy()
	{
		super.onDestroy()
	}

	companion object
	{
		var musicFile: Uri? = null
		var mplayer: MediaPlayer? = null
		var duration = 0
		var layout : View? = null
		var config: Array<String>? = null
		var file : Uri? = null
		var render : Boolean = false
		var context : Activity? = null

		fun stopMusic()
		{
			mplayer!!.stop()
		}

		fun getCurrentPosition(): Int
		{
			return mplayer!!.currentPosition
		}

		fun pauseMusic()
		{
			mplayer!!.pause()
		}

		fun startMusic()
		{
			mplayer!!.start()
		}

		fun musicSeekTo(a: Int)
		{
			mplayer!!.seekTo(a)
		}

		fun mute()
		{
			mplayer!!.setVolume(0f, 0f)
		}

	}
}
